src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -1,207 +1,375 @@
-            }
-        }
-
+        }
+
+        internal async Task ProcessMessageAsync(Message message, CancellationToken cancellationToken)
+        {
+            if (!await _messageProcessor.BeginProcessingMessageAsync(message, cancellationToken))
+            {
+                return;
+            }
+
+            ServiceBusTriggerInput input = ServiceBusTriggerInput.New(message);
+            input.MessageReceiver = Receiver;
+
+            FunctionResult result = await _triggerExecutor.TryExecuteAsync(GetTriggerFunctionData(message, input), cancellationToken);
+            await _messageProcessor.CompleteProcessingMessageAsync(message, result, cancellationToken);
+        }
+
+        internal async Task ProcessSessionMessageAsync(IMessageSession session, Message message, CancellationToken cancellationToken)
+        {
+            _messageSession = session;
+            if (!await _sessionMessageProcessor.BeginProcessingMessageAsync(session, message, cancellationToken))
+            {
+                return;
+            }
+
+            ServiceBusTriggerInput input = ServiceBusTriggerInput.New(message);
+            input.MessageSession = session;
+
+            FunctionResult result = await _triggerExecutor.TryExecuteAsync(GetTriggerFunctionData(message, input), cancellationToken);
+            await _sessionMessageProcessor.CompleteProcessingMessageAsync(session, message, result, cancellationToken);
+        }
+
+        internal void StartMultipleMessageListening(CancellationToken cancellationToken)
+        {
+            SessionClient sessionClient = null;
+            IMessageReceiver receiver = null;
+            if (_isSessionsEnabled)
+            {
+                sessionClient = _messagingProvider.CreateSessionClient(_entityPath, _serviceBusAccount.ConnectionString);
+            }
+            else
+            {
+                receiver = Receiver;
+            }
+
+            Task.Run(async () =>
+            {
+                while (true)
+                {
+                    try
+                    {
+                        if (_receiver == null || cancellationToken.IsCancellationRequested)
+                        {
+                            return;
+                        }
+
+                        if (_isSessionsEnabled)
+                        {
+                            try
+                            {
+                                receiver = await sessionClient.AcceptMessageSessionAsync();
+                            }
+                            catch (ServiceBusTimeoutException)
+                            {
+                                // it's expected if the entity is empty, try next time
+                                continue;
+                            }
+                        }
+
+                        IList<Message> messages = messages = await receiver.ReceiveAsync(_serviceBusOptions.BatchOptions.MaxMessageCount, _serviceBusOptions.BatchOptions.OperationTimeout);

```

Comment:
What happens if the operation timeout expires? Is an exception thrown, or are we just waiting until either MaxMessages are available, otherwise we return 0 messages?

Reply:
1. It returns `null` messages if there are no new messages during "timeout"
2. If there are new message it returns immediately with `messages.Count <= MaxMessages`